arangod/RocksDBEngine/RocksDBCollectionMeta.cpp

```diff
@@ -0,0 +1,419 @@
+
+#include <velocypack/Iterator.h>
+#include <velocypack/velocypack-aliases.h>
+
+using namespace arangodb;
+
+RocksDBCollectionMeta::DocCount::DocCount(VPackSlice const& slice)
+: _committedSeq(0), _added(0), _removed(0), _revisionId(0) {
+  if (!slice.isArray()) {
+    // got a somewhat invalid slice. probably old data from before the key
+    // structure changes
+    return;
+  }
+  
+  VPackArrayIterator array(slice);
+  if (array.valid()) {
+    this->_committedSeq = (*array).getUInt();
+    // versions pre 3.4 stored only a single "count" value
+    // 3.4 and higher store "added" and "removed" seperately
+    this->_added = (*(++array)).getUInt();
+    if (array.size() > 3) {
+      TRI_ASSERT(array.size() == 4);
+      this->_removed = (*(++array)).getUInt();
+    }
+    this->_revisionId = (*(++array)).getUInt();
+  }
+}
+
+void RocksDBCollectionMeta::DocCount::toVelocyPack(VPackBuilder& b) const {
+  b.openArray();
+  b.add(VPackValue(_committedSeq));
+  b.add(VPackValue(_added));
+  b.add(VPackValue(_removed));
+  b.add(VPackValue(_revisionId));
+  b.close();
+}
+
+RocksDBCollectionMeta::RocksDBCollectionMeta()
+  : _count(0,0,0,0) {}
+
+/**
+ * @brief Place a blocker to allow proper commit/serialize semantics
+ *
+ * Should be called immediately prior to internal RocksDB commit. If the
+ * commit succeeds, any inserts/removals should be buffered, then the blocker
+ * removed; otherwise simply remove the blocker.
+ *
+ * @param  trxId The identifier for the active transaction
+ * @param  seq   The sequence number immediately prior to call
+ * @return       May return error if we fail to allocate and place blocker
+ */
+Result RocksDBCollectionMeta::placeBlocker(uint64_t trxId, rocksdb::SequenceNumber seq) {
+  Result res = basics::catchToResult([&]() -> Result {
+    TRI_ASSERT(_blockers.end() == _blockers.find(trxId));
+    TRI_ASSERT(_blockersBySeq.end() ==
+               _blockersBySeq.find(std::make_pair(seq, trxId)));
+    Result res;
+    WRITE_LOCKER(locker, _blockerLock);
+    auto insert = _blockers.emplace(trxId, seq);
+    auto crosslist = _blockersBySeq.emplace(seq, trxId);
+    if (!insert.second || !crosslist.second) {
+      return {TRI_ERROR_INTERNAL};
+    }
+    return {TRI_ERROR_NO_ERROR};
+  });
+  return res;
+}
+
+/**
+ * @brief Removes an existing transaction blocker
+ *
+ * Should be called after transaction abort/rollback, or after buffering any
+ * updates in case of successful commit. If no blocker exists with the
+ * specified transaction identifier, then this will simply do nothing.
+ *
+ * @param trxId Identifier for active transaction (should match input to
+ *              earlier `placeBlocker` call)
+ */
+void RocksDBCollectionMeta::removeBlocker(uint64_t trxId) {
+  WRITE_LOCKER(locker, _blockerLock);
+  auto it = _blockers.find(trxId);
+  if (ADB_LIKELY(_blockers.end() != it)) {
+    auto cross = _blockersBySeq.find(std::make_pair(it->second, it->first));
```

Comment:
you could just use `_blockersBySeq.erase(...)` instead of looking up the value and then deleting it.
the result will be the same.